products/bridge/smart-contracts/contracts/periphery/TokenManagerV2/TokenManagerUpgradeableV2.sol

```diff
@@ -0,0 +1,227 @@
+    }
+
+    function getRemoteTokens(
+        address token,
+        uint remoteChainId
+    ) public view returns (RemoteToken memory) {
+        TokenManagerStorage storage $ = _getTokenManagerStorage();
+        return $.remoteTokens[token][remoteChainId];
+    }
+
+    modifier onlyGateway() {
+        if (msg.sender != address(getGateway())) {
+            revert NotGateway();
+        }
+        _;
+    }
+
+    function __TokenManager_init(address _gateway) internal onlyInitializing {
+        __Ownable_init(msg.sender);
+        _setGateway(_gateway);
+    }
+
+    function _authorizeUpgrade(address) internal virtual override onlyOwner {}
+
+    function _setGateway(address _gateway) internal {
+        TokenManagerStorage storage $ = _getTokenManagerStorage();
+        $.gateway = _gateway;
+    }
+
+    function setGateway(address _gateway) external onlyOwner {
+        _setGateway(_gateway);
+    }
+
+    function _removeToken(address localToken, uint remoteChainId) internal {
+        TokenManagerStorage storage $ = _getTokenManagerStorage();
+        delete $.remoteTokens[localToken][remoteChainId];
+        emit TokenRemoved(localToken, remoteChainId);
+    }
+
+    function _registerToken(
+        address localToken,
+        RemoteToken memory remoteToken
+    ) internal {
+        TokenManagerStorage storage $ = _getTokenManagerStorage();
+        $.remoteTokens[localToken][remoteToken.chainId] = remoteToken;
+        emit TokenRegistered(
+            localToken,
+            remoteToken.token,
+            remoteToken.tokenManager,
+            remoteToken.chainId
+        );
+    }
+
+    // Token Overrides
+    function registerToken(
+        address token,
+        RemoteToken memory remoteToken
+    ) external virtual onlyOwner {
+        _registerToken(token, remoteToken);
+    }
+
+    // V2 New Function
+    function setFees(
+        uint newFees
+    ) external override(ITokenManager, TokenManagerFees) onlyOwner {
+        _setFees(newFees);
+    }
+
+    // V2 New Function
+    function withdrawFees(
+        address payable to
+    ) external override(ITokenManager, TokenManagerFees) onlyOwner {
+        _withdrawFees(to);
+    }
+
+    // V2 New Function
+    function pause() external onlyOwner {
+        _pause();
+    }
+
+    // V2 New Function
+    function unpause() external onlyOwner {
+        _unpause();
+    }
+
+    // TO OVERRIDE – Incoming
+    function _handleTransfer(
+        address token,
+        address from,
+        uint amount
+    ) internal virtual;
+
+    // TO OVERRIDE – Outgoing
+    function _handleAccept(
+        address token,
+        address recipient,
+        uint amount
+    ) internal virtual;
+
+    // V2 Modified: `whenNotPaused` & `checkFees` modifiers, also made payable
+    function transfer(
+        address token,
+        uint remoteChainId,
+        address remoteRecipient,
+        uint amount
+    ) external payable virtual whenNotPaused checkFees {
```

Comment:
Only pauses `transfer` and not accept, because we probably want the remaining accept calls to at least go through before pausing